crates/storage/provider/src/providers/database.rs

```diff
@@ -64,6 +64,12 @@ impl<DB: Database> ShareableDatabase<DB> {
         // +1 as the changeset that we want is the one that was applied after this block.
         block_number += 1;
 
+        // Return an error if requested block is not synced yet
+        let latest = best_block_number(&tx)?.unwrap_or_default();
```

Comment:
this is not the exact location I had in mind, because this now always does another read

ptal at the issue again

Reply:
in the referenced function of BlockchainProvider StateProvider impl